File: downloads/code/LeetCode/Java/Divide-Two-Integers.java
public class Solution {
    public int divide(int dividend, int divisor) {
	if(divisor==0 || (dividend == Integer.MIN_VALUE && divisor==-1))
	    return Integer.MAX_VALUE;
	int sign = (dividend > 0) ^ (divisor > 0) ? -1 : 1;
	long dvd = Math.abs((long)dividend);
	long dvs = Math.abs((long)divisor);
	int ret = 0;
	while(dvd >= dvs){
	    long tmp = dvs;
	    int multiple = 1;
	    while(dvd >= (tmp<<1)){
		tmp <<= 1;
		multiple <<= 1;
	    }
	    dvd -= tmp;
	    ret += multiple;
	}
	return sign==1? ret: -ret;
    }
}

File: downloads/code/LeetCode/Python/Divide-Two-Integers.py
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        max = 2147483647
        if divisor == 0:
            return max
        neg = (dividend < 0 and divisor > 0) or (dividend > 0 and divisor < 0)
        a = abs(dividend)
        b = abs(divisor)

        ret = 0
        shift = 31

        while shift >= 0:
            if a >= (b<<shift):
                a -= b<<shift
                ret += 1<<shift
            shift -= 1

        if neg:
            ret = -ret
        if ret >= max:
            return max
        return ret
